### ISSUE TITLE
chained comparison

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#16317]
Migrated from [rt.perl.org#132607](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132607) (status was 'open')

Searchable as RT132607$

**Comment by p5pRT** [comment]
### From zefram@fysh.org

#### Created by zefram@fysh.org



It would be nice for Perl to understand chained comparisons such as "$a
le $b lt $c"\, which would mean "$a le $b && $b lt $c" but with only a
single evaluation of the middle comparand\.  The comparison operators are
currently non\-associative\, so most of what ought to be chained comparisons
are currently syntax errors\.  Introducing chained comparisons would not
change the meaning of any currently\-legal expression\.

One wrinkle&#8203;: we currently have two levels of precedence for comparison
operators\, one for "lt" et al\, and a looser one for "eq" et al\.  So "$a lt
$b eq $c" is currently legal\.  This is imitating the C precedence levels\,
I think mainly so that "\!=" can be used to mean "xor"\.    It's already a
bit of a gotcha\, and we'd have to warn that things like that don't make
chained comparisons\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=wishlist

Site configuration information for perl 5.27.6:

Configured by zefram at Tue Nov 21 05:42:59 GMT 2017.

Summary of my perl5 (revision 5 version 27 subversion 6) configuration:
   
  Platform:
    osname=linux
    osvers=3.16.0-4-amd64
    archname=x86_64-linux-thread-multi
    uname='linux barba.rous.org 3.16.0-4-amd64 #1 smp debian 3.16.43-2+deb8u2 (2017-06-26) x86_64 gnulinux '
    config_args='-des -Dprefix=/home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52 -Duselargefiles -Dusethreads -Uafs -Ud_csh -Uusesfio -Uusenm -Duseshrplib -Dusedevel -Uversiononly -Ui_db'
    hint=recommended
    useposix=true
    d_sigaction=define
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=define
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cc'
    ccflags ='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -D_FORTIFY_SOURCE=2'
    optimize='-O2'
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fwrapv -fno-strict-aliasing -pipe -fstack-protector-strong -I/usr/local/include'
    ccversion=''
    gccversion='4.9.2'
    gccosandvers=''
    intsize=4
    longsize=8
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=define
    longlongsize=8
    d_longdbl=define
    longdblsize=16
    longdblkind=3
    ivtype='long'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='off_t'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='cc'
    ldflags =' -fstack-protector-strong -L/usr/local/lib'
    libpth=/usr/local/lib /usr/lib/gcc/x86_64-linux-gnu/4.9/include-fixed /usr/include/x86_64-linux-gnu /usr/lib /lib/x86_64-linux-gnu /lib/../lib /usr/lib/x86_64-linux-gnu /usr/lib/../lib /lib
    libs=-lpthread -lnsl -ldb -ldl -lm -lcrypt -lutil -lc
    perllibs=-lpthread -lnsl -ldl -lm -lcrypt -lutil -lc
    libc=libc-2.19.so
    so=so
    useshrplib=true
    libperl=libperl.so
    gnulibc_version='2.19'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs
    dlext=so
    d_dlsymun=undef
    ccdlflags='-Wl,-E -Wl,-rpath,/home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52/lib/5.27.6/x86_64-linux-thread-multi/CORE'
    cccdlflags='-fPIC'
    lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector-strong'



@INC for perl 5.27.6:
    /home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52/lib/site_perl/5.27.6/x86_64-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52/lib/site_perl/5.27.6
    /home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52/lib/5.27.6/x86_64-linux-thread-multi
    /home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52/lib/5.27.6


Environment for perl 5.27.6:
    HOME=/home/zefram
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/zefram/usr/perl/perl_install/perl-5.27.6-i64-f52/bin:/home/zefram/usr/perl/util:/home/zefram/pub/x86_64-unknown-linux-gnu/bin:/home/zefram/pub/common/bin:/usr/bin:/bin:/usr/local/bin:/usr/games
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

I so look forward to this\, both as a patch i can learn from and as a
feature I would enjoy\. \+\+Zefram\.

Yves

On 19 Dec 2017 7&#8203;:38 am\, "Zefram" \<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> \# New Ticket Created by  Zefram
> \# Please include the string&#8203;:  \[perl \#132607\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=132607 >
>
>
>
> This is a bug report for perl from zefram@&#8203;fysh\.org\,
> generated with the help of perlbug 1\.41 running under perl 5\.27\.6\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> It would be nice for Perl to understand chained comparisons such as "$a
> le $b lt $c"\, which would mean "$a le $b && $b lt $c" but with only a
> single evaluation of the middle comparand\.  The comparison operators are
> currently non\-associative\, so most of what ought to be chained comparisons
> are currently syntax errors\.  Introducing chained comparisons would not
> change the meaning of any currently\-legal expression\.
>
> One wrinkle&#8203;: we currently have two levels of precedence for comparison
> operators\, one for "lt" et al\, and a looser one for "eq" et al\.  So "$a lt
> $b eq $c" is currently legal\.  This is imitating the C precedence levels\,
> I think mainly so that "\!=" can be used to mean "xor"\.  It's already a
> bit of a gotcha\, and we'd have to warn that things like that don't make
> chained comparisons\.
>
> \[Please do not change anything below this line\]
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \-\-\-
> Flags&#8203;:
>     category=core
>     severity=wishlist
> \-\-\-
> Site configuration information for perl 5\.27\.6&#8203;:
>
> Configured by zefram at Tue Nov 21 05&#8203;:42&#8203;:59 GMT 2017\.
>
> Summary of my perl5 \(revision 5 version 27 subversion 6\) configuration&#8203;:
>
>   Platform&#8203;:
>     osname=linux
>     osvers=3\.16\.0\-4\-amd64
>     archname=x86\_64\-linux\-thread\-multi
>     uname='linux barba\.rous\.org 3\.16\.0\-4\-amd64 \#1 smp debian
> 3\.16\.43\-2\+deb8u2 \(2017\-06\-26\) x86\_64 gnulinux '
>     config\_args='\-des \-Dprefix=/home/zefram/usr/
> perl/perl\_install/perl\-5\.27\.6\-i64\-f52 \-Duselargefiles \-Dusethreads \-Uafs
> \-Ud\_csh \-Uusesfio \-Uusenm \-Duseshrplib \-Dusedevel \-Uversiononly \-Ui\_db'
>     hint=recommended
>     useposix=true
>     d\_sigaction=define
>     useithreads=define
>     usemultiplicity=define
>     use64bitint=define
>     use64bitall=define
>     uselongdouble=undef
>     usemymalloc=n
>     default\_inc\_excludes\_dot=define
>     bincompat5005=undef
>   Compiler&#8203;:
>     cc='cc'
>     ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing
> \-pipe \-fstack\-protector\-strong \-I/usr/local/include \-D\_LARGEFILE\_SOURCE
> \-D\_FILE\_OFFSET\_BITS=64 \-D\_FORTIFY\_SOURCE=2'
>     optimize='\-O2'
>     cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-fwrapv \-fno\-strict\-aliasing
> \-pipe \-fstack\-protector\-strong \-I/usr/local/include'
>     ccversion=''
>     gccversion='4\.9\.2'
>     gccosandvers=''
>     intsize=4
>     longsize=8
>     ptrsize=8
>     doublesize=8
>     byteorder=12345678
>     doublekind=3
>     d\_longlong=define
>     longlongsize=8
>     d\_longdbl=define
>     longdblsize=16
>     longdblkind=3
>     ivtype='long'
>     ivsize=8
>     nvtype='double'
>     nvsize=8
>     Off\_t='off\_t'
>     lseeksize=8
>     alignbytes=8
>     prototype=define
>   Linker and Libraries&#8203;:
>     ld='cc'
>     ldflags =' \-fstack\-protector\-strong \-L/usr/local/lib'
>     libpth=/usr/local/lib /usr/lib/gcc/x86\_64\-linux\-gnu/4\.9/include\-fixed
> /usr/include/x86\_64\-linux\-gnu /usr/lib /lib/x86\_64\-linux\-gnu /lib/\.\./lib
> /usr/lib/x86\_64\-linux\-gnu /usr/lib/\.\./lib /lib
>     libs=\-lpthread \-lnsl \-ldb \-ldl \-lm \-lcrypt \-lutil \-lc
>     perllibs=\-lpthread \-lnsl \-ldl \-lm \-lcrypt \-lutil \-lc
>     libc=libc\-2\.19\.so
>     so=so
>     useshrplib=true
>     libperl=libperl\.so
>     gnulibc\_version='2\.19'
>   Dynamic Linking&#8203;:
>     dlsrc=dl\_dlopen\.xs
>     dlext=so
>     d\_dlsymun=undef
>     ccdlflags='\-Wl\,\-E \-Wl\,\-rpath\,/home/zefram/usr/
> perl/perl\_install/perl\-5\.27\.6\-i64\-f52/lib/5\.27\.6/x86\_64\-
> linux\-thread\-multi/CORE'
>     cccdlflags='\-fPIC'
>     lddlflags='\-shared \-O2 \-L/usr/local/lib \-fstack\-protector\-strong'
>
>
> \-\-\-
> @&#8203;INC for perl 5\.27\.6&#8203;:
>     /home/zefram/usr/perl/perl\_install/perl\-5\.27\.6\-i64\-f52/
> lib/site\_perl/5\.27\.6/x86\_64\-linux\-thread\-multi
>     /home/zefram/usr/perl/perl\_install/perl\-5\.27\.6\-i64\-f52/
> lib/site\_perl/5\.27\.6
>     /home/zefram/usr/perl/perl\_install/perl\-5\.27\.6\-i64\-f52/
> lib/5\.27\.6/x86\_64\-linux\-thread\-multi
>     /home/zefram/usr/perl/perl\_install/perl\-5\.27\.6\-i64\-f52/lib/5\.27\.6
>
> \-\-\-
> Environment for perl 5\.27\.6&#8203;:
>     HOME=/home/zefram
>     LANG \(unset\)
>     LANGUAGE \(unset\)
>     LD\_LIBRARY\_PATH \(unset\)
>     LOGDIR \(unset\)
>     PATH=/home/zefram/usr/perl/perl\_install/perl\-5\.27\.6\-i64\-
> f52/bin&#8203;:/home/zefram/usr/perl/util&#8203;:/home/zefram/pub/x86\_64\-
> unknown\-linux\-gnu/bin&#8203;:/home/zefram/pub/common/bin&#8203;:/usr/
> bin&#8203;:/bin&#8203;:/usr/local/bin&#8203;:/usr/games
>     PERL\_BADLANG \(unset\)
>     SHELL=/usr/bin/zsh
>
>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @leonerd

On Tue\, 19 Dec 2017 09&#8203;:10&#8203;:57 \+0100
demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:

> I so look forward to this\, both as a patch i can learn from and as a
> feature I would enjoy\. \+\+Zefram\.

\+1

I'd be happy to accept some new warnings\, in exchange for being able to
e\.g\.&#8203;:

&nbsp;  grep \{ $min \<= $\_ \<= $max \} @&#8203;numbers

\-\- 
Paul "LeoNerd" Evans

leonerd@&#8203;leonerd\.org\.uk      |  https://metacpan.org/author/PEVANS
http&#8203;://www\.leonerd\.org\.uk/  |  https://www.tindie.com/stores/leonerd/

**Comment by p5pRT** [comment]
### From @davidnicol

how would this be done without looking ahead more than one parse\-time
element?

Perhaps a "Comparison Expression" type would be introduced to the grammar\.
Except the fact of the current syntax error implies there is already a
place the new thing could go\.

Expressions in the middle should only get evaluated once\, and later ones
should short\-circuit out&#8203;:

&nbsp;        $min \<= &lookup1 \< &lookup2 \<= $max

should be equivalent to

&nbsp;        do \{  my $tmp1 = &lookup1;  $min \<= $tmp1 and do \{  my $tmp2 =
&lookup2; $tmp1 \< $tmp2 and $tmp2 \<= $max  \}   \}

rather than the double\-calling

&nbsp;        \( $min \<= &lookup1 and &lookup1 \< &lookup2 and  &lookup2 \<= $max \)

for any expression &lookupN more complicated than an unblessed scalar\.


I think I'm willing to write a test case for this if it would help\.



On Wed\, Dec 20\, 2017 at 10&#8203;:39 AM\, Paul "LeoNerd" Evans \<
leonerd@&#8203;leonerd\.org\.uk> wrote&#8203;:

> On Tue\, 19 Dec 2017 09&#8203;:10&#8203;:57 \+0100
> demerphq \<demerphq@&#8203;gmail\.com> wrote&#8203;:
>
> > I so look forward to this\, both as a patch i can learn from and as a
> > feature I would enjoy\. \+\+Zefram\.
>
> \+1
>
> I'd be happy to accept some new warnings\, in exchange for being able to
> e\.g\.&#8203;:
>
>   grep \{ $min \<= $\_ \<= $max \} @&#8203;numbers
>
> \-\-
> Paul "LeoNerd" Evans
>
> leonerd@&#8203;leonerd\.org\.uk      |  https://metacpan.org/author/PEVANS
> http&#8203;://www\.leonerd\.org\.uk/  |  https://www.tindie.com/stores/leonerd/
>



\-\- 
“no man should be compelled to do what the laws do not require; nor to
refrain from acts which the laws permit\.” Calder v\. Bull \(U\.S\. 1798\)

**Comment by xenu** [comment]
This feature was implemented in perl 5.32, see 02b85d3dab092d678cfc958a2dc252405333ed25 and the corresponding [perldelta entry](https://perldoc.pl/5.32.0/perldelta#Chained-comparisons-capability).
